plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange.hs

```diff
@@ -75,27 +74,45 @@ selectionRangeHandler ide _ SelectionRangeParams{..} = do
     positions :: [Position]
     List positions = _positions
 
-    logger = ideLogger ide
-
 getSelectionRanges :: NormalizedFilePath -> [Position] -> ExceptT String IdeAction [SelectionRange]
 getSelectionRanges file positions = do
-    (HAR{hieAst, refMap}, positionMapping) <- maybeToExceptT "fail to get hie ast" $ useE GetHieAst file
-    -- 'positionMapping' should be applied to the input positions before using them
+    (codeRange, positionMapping) <- maybeToExceptT "fail to get code range" $ useE GetCodeRange file
+    -- 'positionMapping' should be appied to the input before using them
     positions' <- maybeToExceptT "fail to apply position mapping to input positions" . MaybeT . pure $
         traverse (fromCurrentPosition positionMapping) positions
 
-    ast <- maybeToExceptT "fail to get ast for current file" . MaybeT . pure $
-        -- in GHC 9, the 'FastString' in 'HieASTs' is replaced by a newtype wrapper around 'LexicalFastString'
-        -- so we use 'coerce' to make it work in both GHC 8 and 9
-        getAsts hieAst Map.!? (coerce . mkFastString . fromNormalizedFilePath) file
-
-    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
-    let selectionRanges = findSelectionRangesByPositions (astPathsLeafToRoot ast') positions'
+    let selectionRanges = flip fmap positions' $ \pos ->
+            -- codeRange doesn't cover all portions of text in the file, so we need a default value
+            let defaultSelectionRange = SelectionRange (Range pos pos) Nothing
+             in reverseSelectionRange . fromMaybe defaultSelectionRange . findPosition' pos $ codeRange
 
     -- 'positionMapping' should be applied to the output ranges before returning them
     maybeToExceptT "fail to apply position mapping to output positions" . MaybeT . pure $
          traverse (toCurrentSelectionRange positionMapping) selectionRanges
 
+-- Find 'Position' in 'CodeRange'. Producing an inverse 'SelectionRange'
+findPosition' :: Position -> CodeRange -> Maybe SelectionRange
+findPosition' pos (CodeRange range children) =
+    if positionInRange pos range
+    then Just $ case mapMaybe (findPosition' pos) children of
+            [childSelectionRange] -> SelectionRange range (Just childSelectionRange)
+            _                     -> SelectionRange range Nothing
+    else Nothing
+
+-- Reverse 'SelectionRange'. Just like 'reverse' for list.
+reverseSelectionRange :: SelectionRange -> SelectionRange
+reverseSelectionRange = go (SelectionRange invalidRange Nothing)
+  where
+    go :: SelectionRange -> SelectionRange -> SelectionRange
+    go acc (SelectionRange r Nothing) = SelectionRange r (checkRange acc)
+    go acc (SelectionRange r (Just parent)) = go (SelectionRange r (checkRange acc)) parent
+
+    checkRange :: SelectionRange -> Maybe SelectionRange
```

Comment:
what's going on here? why does this happen?

Reply:
This is basically trying to reverse a non-empty list (`SelectionRange` is basically a non-empty list). Unlike reversing an ordinary list, I can't find an appropriate initial value for `go`. Maybe I should use `Maybe`